[RHCLOUD-21533]Follow up PR to Liveness and Readiness Probe for Background Worker

Ellen-Yi-Dong opened the pull request:
What: Adding readiness and liveness probe for background worker and added health check to background worker 

This is a followup PR from this PR: https://github.com/RedHatInsights/sources-api-go/pull/599

JIRA: [RHCLOUD-21533](https://issues.redhat.com/browse/RHCLOUD-21533)

lindgrenj6 commented on deploy/clowdapp.yaml:
```diff
@@ -64,6 +64,17 @@ objects:
+              path: /health
+              port: 8000
+            initialDelaySeconds: 10
+            periodSeconds: 30
```
This just needs indented 2 spaces to the left!

lindgrenj6 commented on jobs/worker.go:
```diff
@@ -57,8 +59,13 @@ func Run(shutdown chan struct{}) {
 	}()
 
+	//create struct for healthcheck struct and run the healthcheck thread
+	healthCheck := healthCheckStruct{timeStamp: time.Now(), timeChannel: make(chan time.Time, 10)}
```
We should probably rename the struct to not include `struct` in the name. What do you think of `BackgroundWorkerHealthChecker`?

lindgrenj6 commented on jobs/worker.go:
```diff
@@ -91,3 +98,45 @@ func RunJobNow(j Job) {
+	go hc.healthCheckConsumer()
+
+	//create new GET request through echo with the same path name "/health"
+	e := echo.New()
```
Can we also tack
```go
	e.HideBanner = true
	e.HidePort = true
```
After this? So we don't get the ASCII art output to keep our logs pure JSON.

lindgrenj6 commented on jobs/worker.go:
```diff
@@ -91,3 +98,45 @@ func RunJobNow(j Job) {
+			return c.String(http.StatusInternalServerError, "500")
+		}
+		return c.String(http.StatusOK, "OK")
+
```
I think this empty line can go after the bracket.

lindgrenj6 commented on jobs/worker.go:
```diff
@@ -91,3 +98,45 @@ func RunJobNow(j Job) {
+
+// healthcheckConsumer runs a loop against the healthcheck time channel and updates the shared timestamp with the last time we received a response
+func (hc *healthCheckStruct) healthCheckConsumer() {
+	for {
```
We can probably change this to `for range hc.timeChannel` since we don't have to manually pull from the channel anymore.

lindgrenj6 commented on jobs/worker.go:
```diff
@@ -91,3 +98,45 @@ func RunJobNow(j Job) {
+func (hc *healthCheckStruct) healthcheckLoop() {
+	//every 30 seconds, send a ping request to redis server. If ping request returns a <nil> error, then update the time
+	for range time.NewTicker(30 * time.Second).C {
+		redisErr := redis.Client.Ping(context.Background()).Err()
```
Can we rename this to `err`? All err's should be the same name.

lindgrenj6 commented on redis/client.go:
```diff
@@ -2,7 +2,6 @@ package redis
 
 import (
 	"fmt"
-
```
I think both of these changes could be removed - since we didn't have to touch this file.

lindgrenj6 reviewed (changes requested):
Nice work! I just have a bunch of teeny style changes :grin:

Ellen-Yi-Dong committed:
Added readiness and liveness probe for background worker within clowdapp.yml

Created health check within jobs/worker.go

Added style changes to worker.go file

Added indentations to the left in  clowdapp.yaml file.

lindgrenj6 commented on deploy/clowdapp.yaml:
```diff
@@ -64,6 +64,17 @@ objects:
+            path: /health
+            port: 8000
+          initialDelaySeconds: 10
+        periodSeconds: 30
```
```suggestion
        readinessProbe:
          httpGet:
            path: /health
            port: 8000
          initialDelaySeconds: 1
        livenessProbe:
          httpGet:
            path: /health
            port: 8000
          initialDelaySeconds: 10
          periodSeconds: 30
```

sorry, this is super complex. this _should_ work. this way you can click apply and it should take the diff for you.

lindgrenj6 commented on jobs/worker.go:
```diff
@@ -91,3 +99,47 @@ func RunJobNow(j Job) {
+		err := redis.Client.Ping(context.Background()).Err()
+		if err == nil {
+			hc.timeChannel <- time.Now()
+		}
```
I should have mentioned this earlier - can we add an `else` here that logs a failure to hit redis?

lindgrenj6 commented on jobs/worker.go:
```diff
@@ -91,3 +99,47 @@ func RunJobNow(j Job) {
+func (hc *BackgroundWorkerHealthChecker) healthCheckConsumer() {
+	for range hc.timeChannel {
+		hc.timeStamp = time.Now()
+		fmt.Println("timeStamp", hc.timeStamp)
```
We probably don't need this print anymore.

lindgrenj6 reviewed (commented):
Getting closer!

Ellen-Yi-Dong committed:
Update deploy/clowdapp.yaml

Co-authored-by: Jacob Lindgren <lindgrenj6@gmail.com>

Ellen-Yi-Dong committed:
Added else statement within healthcheckLoop()

Deleted unnecessary comment

lindgrenj6 commented on jobs/worker.go:
```diff
@@ -91,3 +98,48 @@ func RunJobNow(j Job) {
+		if err == nil {
+			hc.timeChannel <- time.Now()
+		} else {
+			l.Log.Fatalf("Failed to hit redis: %v", err)
```
This should be a `Warnf`, not a `Fatalf`. We don't want the process to exit on failure, the k8s liveness probe will do it for us. 

This will also allow the log to be uploaded to cloudwatch so we can see it in kibana and know "huh we started failing to hit redis at x:xx pm"

lindgrenj6 commented on jobs/worker.go:
```diff
@@ -91,3 +98,48 @@ func RunJobNow(j Job) {
+	e.GET("/health", func(c echo.Context) error {
+		//health should return a 500 if there hasn't been a response from ping request within 30 seconds
+		if hc.timeStamp.Before(time.Now().Add(-30 * time.Second)) {
+			return c.String(http.StatusInternalServerError, "500")
```
And one last comment here - can we change the `"500"` message to `"Failed to hit redis for more than 30 seconds"`?

lindgrenj6 reviewed (commented):
Last comments!!

Ellen-Yi-Dong committed:
Change error messages within healthchecker() and healthcheckLoop()

lindgrenj6 reviewed (approved):
Great work!
